### DISCUSSION TITLE
[A-L] (14 marzo 2024) Esercizio 1

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#17] [Esercitazioni A-L]
In un grafo non diretto e connesso $G = (V, E)$ un vertice $v$ si dice di **articolazione** _(**cutvertex** in inglese)_ se $G − v$, il grafo ottenuto da $G$ rimuovendo $v$ e tutti gli archi ad esso incidenti, non è connesso. Modificare l’algoritmo della ricerca in profondità in maniera da poter ottenere tutti e soli i vertici di articolazione di un grafo connesso; è possibile fare questa modifica in modo che il controllo avvenga in $\theta(|V|+|E|)$?

Come si può ulteriormente modificare l’algoritmo per ottenere tutti i vertici di articolazione di un grafo non necessariamente connesso _(dove in questo caso il criterio di un vertice di articolazione è quello di disconnettere la componente connessa che lo contiene)_?

**Comment by LeoCosta02** [comment]
main_articolazione(Graph: G){
t: int[n]
c: int=1
articolazione: list
x: vertice a caso
children : int=0
 while( esiste x.adiacente AND children < 2) //essendo x la radice se ha almeno 2 archi incidenti: articolazione
	 children ++
 if(children >=2)
	 articolazione.add(x)
 
 t[x]=c
 y=x.adiacente // un adiacente di x 
 dfs_articolazione(G,x,y,t,articolazione,c)
 return articolazione	 


}



dfs_articolazione(graph:G ,node_padre: z node: x , t[] , articolazione : list, c: int){

 c++
 t[x]=c
 back=c

	for each y adiacente di x {
		if(t[y]==0){
		b=dfs_articolazione(G,x,y,t,articolazione,c)
		back=min(b,back)
		}
			else{
				if(y!=z){
			    back=min(t[y],back)
				}
			}
	}
		if(t[x]==back AND x != z)
			articolazione.add(x)
		
		
		return back 
}

**Comment by daromacarlo** [comment]
![Screenshot from 2024-03-18 15-35-05](https://github.com/sapienzastudentsnetwork/progalg2324/assets/135441131/51e2d56e-91a9-45e9-94fb-db29308be6ca)
![Screenshot from 2024-03-18 15-35-25](https://github.com/sapienzastudentsnetwork/progalg2324/assets/135441131/a49d75a7-5199-4290-b392-0eaf092b3a18)
![Screenshot from 2024-03-18 15-35-32](https://github.com/sapienzastudentsnetwork/progalg2324/assets/135441131/0f5063a9-0fc0-4f18-86e0-5a16225b718b)

**Comment by cornflexxx** [comment]
![ponte](https://github.com/sapienzastudentsnetwork/progalg2324/assets/121803863/d095f703-2d4c-4e62-a1fa-0d0096211f45)
![articol1](https://github.com/sapienzastudentsnetwork/progalg2324/assets/121803863/ca94407a-856d-4eb9-9717-ce94605c8dd4)

Consideriamo l'albero di visita creato dalla $DFS$ su un grafo non diretto , allora le articolazioni possono essere trovate con i seguenti criteri:

- Un vertice $u$ con $deg(u)\le 1$ non è mai articolazione, se pariliamo di albero di visita allora significa che il nodo non ha figli.
- Se  $deg(u)\ge2$, allora sarà un'articolazione (_fig.1/fig.2_)  se possiede almeno un figlio e non ha archi all'indietro che partono dai suoi sotto alberi e lo superano.

Con questa distinzione però non riusciamo ad accorgerci se la radice non è un'articolazione, quindi possiamo porre un controllo sulla radice e vedere che questa è un'articolazione $\Longleftrightarrow$ ha almeno due figli.

Il seguente algoritmo trova tutti le articolazioni di una componente e le salva nell'array `articolazioni`.

```

DFS_articolazione(v: vertice di partenza, articolazioni: lista vuota, tempo_visita, contatore, visitato):
    contatore = contatore + 1
    figli = 0
    visitato[v] = 1
    tempo_visita[v] = contatore
    arco_indietro = contatore
    for u in v.adiacenti:
    	if visitato[u] == 0:
    		arco = DFS(u, articolazioni, tempo_visita, contatore, visitato)
    		figli = figli + 1
    		arco_indietro = min(arco, arco_indietro)
    	else if tempo_visita[u] == contatore -1:
    		continue
    	else:
    		arco_indietro = min(tempo_visita[u],arco_indietro)
    if tempo_visita == 1:
    	if figli >= 2:
    		articolazioni.add(v)
    elif (arco_indietro == tempo_visita) and (figli >= 1):
    	articolazione.add(v)
    return arco_indietro
    

```

**Comment by Frisbon** [comment]
Ragionamento analogo a @CarloDaRomadev _(quel bellissimo uomo sul cavallo mammamia me lo farei)_

![image_2024-06-14_10-07-14](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/5cf07248-d29b-49e4-97fd-81f6078ae8a7)

**Comment by daromacarlo** [comment]
@Frisbon SLURP
